src/Language/Fixpoint/Types/Names.hs

```diff
@@ -161,63 +165,114 @@ type SafeText = T.Text
 --
 --   where i is a unique integer (for each text)
 
-data Symbol
+data FixSymbol
   = S { _symbolId      :: !Id
       , symbolRaw     :: !T.Text
       , symbolEncoded :: !T.Text
       } deriving (Data, Typeable, Generic)
 
-instance Eq Symbol where
+data AbstractSymbol s
+  = PS { abstractSymbol :: s
+       , abstractSymbolEncoded :: !T.Text
+       } deriving (Data, Typeable, Generic)
+instance Eq s => Eq (AbstractSymbol s) where
+  s1 == s2 = abstractSymbol s1 == abstractSymbol s2
+instance Show s => Show (AbstractSymbol s) where
+  show as = show (abstractSymbol as)
+
+
+instance (Ord s) => Ord (AbstractSymbol s) where
+  compare s1 s2 = compare (abstractSymbol s1) (abstractSymbol s2)
+
+
+data Symbol s
+  = FS FixSymbol
+  -- | 's' represents an abstract symbol from the source language.
+  | AS (AbstractSymbol s)
+  deriving (Data, Typeable, Generic)
+deriving instance (Eq s) => Eq (Symbol s)
+deriving instance (Ord s) => Ord (Symbol s)
+
+instance (Show s) => Show (Symbol s) where
+  show (FS s) = show s
+  show (AS as) = show as
+
+-- instance IsString (Symbol s) where
+--   fromString = FS . fromString
+
+instance Eq FixSymbol where
   S i _ _ == S j _ _ = i == j
 
-instance Ord Symbol where
+instance Ord FixSymbol where
   -- compare (S i _ _) (S j _ _) = compare i j
   -- compare s1 s2 = compare (symbolString s1) (symbolString s2)
   compare s1 s2 = compare (symbolText s1) (symbolText s2)
 
-instance Interned Symbol where
-  type Uninterned Symbol = T.Text
-  newtype Description Symbol = DT T.Text deriving (Eq)
+instance Interned FixSymbol where
+  type Uninterned FixSymbol = T.Text
+  newtype Description FixSymbol = DT T.Text deriving (Eq)
   describe     = DT
   identify i t = S i t (encode t)
   cache        = sCache
 
-instance Uninternable Symbol where
+instance Uninternable FixSymbol where
   unintern (S _ t _) = t
 
-instance Hashable (Description Symbol) where
+instance Hashable (Description FixSymbol) where
   hashWithSalt s (DT t) = hashWithSalt s t
 
-instance Hashable Symbol where
+instance Hashable FixSymbol where
   -- NOTE: hash based on original text rather than id
   hashWithSalt s (S _ t _) = hashWithSalt s t
 
-instance NFData Symbol where
+
+instance Hashable s => Hashable (Symbol s) where
+  hashWithSalt s sym = hashWithSalt s (toPolynomial sym)
+    where toPolynomial :: Symbol s -> Either FixSymbol (s, T.Text)
+          toPolynomial (AS (PS {abstractSymbolEncoded=encoded, abstractSymbol=as})) = Right (as, encoded)
```

Comment:
You shouldn't need to include `encoded` if we have the invariant that `abstractSymbolEncoded` is a function of `abstractSymbol`. We should add this invariant in the comments of `AbstractSymbol`.